### PR TITLE
feat: inline-reclaim curated tail HOFs at capture-safe call sites (callback specialization A)

### PR DESCRIPTION
**Pull request by aalpar**
## What

At a call site like `(for-each cb xs)` or `(fold + 0 xs)` where the callback is **provably capture-safe**, inline the HOF's single-sequence loop with the callback substituted, so the inlined self-tail loop **reclaims its env frame** instead of leaking ~2 frames/element. Six curated tail HOFs are covered: `for-each`, `vector-map`, `vector-for-each`, `string-map`, `string-for-each`, `fold`.

This realizes the "next lever is allocation reduction" the parallel-scaling arc identified — the stdlib HOFs are the dominant remaining env-frame allocation source.

## Mechanism (no new ABI, no new opcode)

Reuses the shipped `tryInlineCall` synthetic-let substrate:
1. **Stamp (D3)** — a curated `InlineHOF{callbackParamIndex}` capability on the HOF binding's `BindingMeta`, applied post-bootstrap (sealed-base HOFs) or on the import path (`fold`, srfi/1). Curated, not auto-detected.
2. **Callback proof (D2)** — `CallbackIsCaptureSafe`: a symbol resolving to `IsCaptureSafe() && IsStable()`, or a lambda literal passing `ProcedureBodyIsCaptureSafe`. Else → the real HOF.
3. **Unify** — always synthetic-let-bind the callback (symbol or lambda) to the template's first param, stamped `CaptureSafe`. Each template is a validated Scheme loop transcribed from the HOF's bootstrap clause; it inlines via `CompileValidatedLet`, and the self-tail loop's `bodyCalleesAllCaptureSafe` then passes → `OpSelfTailCall` → frame reclaim. `inlineHOFTemplateSource` is the single source of truth (6 templates).

## Soundness (non-negotiable)

Inline the reclaiming form ONLY when the callback is provably capture-safe; uncertain → the real, capturable HOF. A false positive (reclaiming a frame across a capturing callback) is use-after-release — the failure that sank three runtime recyclers. Enforced and tested: a `call/cc` callback is not capture-safe → falls through to the real HOF with the result/traversal intact (one test per shape). `fold` is the one HOF in a mutable imported library; verified that `(import (srfi 1))` marks it `Imported → IsStable` under **both** the immutable default and `WithMutableTopLevel` (R7RS forbids `set!` on imports), so inlining its body cannot go stale. `make test-race` clean.

## Measured payoff

- **Single-thread reclaim** — allocation slope **2.000 → ~0 allocs/element** (the env frame) for every HOF. Strings use a differential measurement: `string-ref`/`string-set!` rebuild the immutable Go-string backing per access (string-map is O(n²)), so the env-frame signal is isolated by subtracting a leaking `(car (list cb))` arm.
- **Parallel (P5, real stdlib `fold`, weak-scaling SRFI-18 sweep, M4 Max)** — **27× allocation cut** (Leaky 32.97 → Reclaim 1.227 MiB/op at P1, deterministic ±0%); per-run advantage **grows with thread count, 1.86× (1T) → 2.38× (8T)**; Reclaim out-scales Leaky at every P. The parallel-scaling plateau is raised, not broken (absolute effective-parallelism rolls off past the P-cores — the inherent result allocation is the next ceiling).

## Test plan
- 14 P6 tests (reclaim slope + inline-vs-real equivalence + `call/cc` soundness per HOF) + P3's `for-each` suite + the D2 callback-proof table + the kept parallel scaling bench (`hof_parallel_scaling_bench_test.go`, deterministic B/op regression guard).
- `make ci` green · `make test-race` green · `make lint` 0 issues · `make covercheck` 41/41 ≥80%.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pkg/machine/compilation/library_bindings.go [hunk 1]
```diff
@@ -62,6 +69,7 @@ func markBindingImported(target, source *environment.Binding) {
 	if source != nil {
 		m.CaptureSafe = source.IsCaptureSafe()
 	}
+	stampInlineHOF(target, exportName)
```
[the remark this answers is not in the document]
Comment (aalpar) [reply pkg/machine/compilation/library_bindings.go]: Fixed in 855d7c31. The import path now stamps only **import-gated** HOFs (`fold`) via `stampImportedInlineHOF`; the sealed-base HOFs (`for-each`, `vector-map`, `string-map`, …) are stamped only at their real home (`StampInlineHOFs`). So a re-export of a sealed-base name — SRFI-13's `string-map` — is no longer stamped, and `(string-map f s)` after `(import (srfi 13))` falls through to SRFI-13's procedure instead of inlining the R7RS template. Pinned by `TestInlineHOFStamp_Srfi13StringMapNotStamped`. (The two parallel maps were also consolidated into one `inlineHOFSpecs` carrying an `importGated` flag, so the distinction is data, not two code paths.)

File: pkg/machine/compilation/compile_call.go [hunk 2]
```diff
@@ -238,3 +246,78 @@ func (p *CompileTimeContinuation) tryInlineCall(
...
+	}
+
+	b := p.env.GetBinding(sym.Symbol.Sym, sym.Symbol.Scopes())
+	if b == nil {
+		return false, nil
+	}
+	cbIdx := b.InlineHOFParam()
+	if cbIdx < 0 {
+		return false, nil
+	}
```
[the remark this answers is not in the document]
Comment (aalpar) [reply pkg/machine/compilation/compile_call.go:284]: Investigated; not adding the gate — the staleness it guards is unreachable, verified two ways. (1) A user `(define fold …)` after import resolves at compile time to a fresh **unstamped** runtime binding (`InlineHOFParam() == -1`, handled by the existing `cbIdx < 0` path), not the stamped import, so the redefinition wins (returns the new value, not a stale inline) — confirmed identical with and without the proposed gate. (2) `(set! fold …)` on an imported binding is rejected (`cannot mutate imported binding`), even under `WithMutableTopLevel`. Structurally, both stamp seams stamp only non-rebindable bindings: `StampInlineHOFs` sweeps the immutable sealed base, and `stampImportedInlineHOF` stamps imported bindings (`Imported ⇒ IsStable`). So a stamped binding reaching `tryInlineHOFCall` is always `IsStable()`; the gate would be a dead branch. Documented this invariant at the `cbIdx < 0` check in 855d7c31 instead.

**Comment by aalpar** [comment]
## Review addressed — Copilot + crosscheck (5 lenses), 2 fix commits

Both review streams aggregated; fixes in `855d7c31` (production) and `df76f4f2` (tests). `make ci` green, `make lint` 0, `make covercheck` 41/41 ≥80%, `go test -race` clean.

### Copilot inline comments
- **`library_bindings.go:72` (import stamp by export-name) — fixed.** The import path now stamps only import-gated HOFs (`fold`); SRFI-13's `string-map` re-export is no longer stamped. Regression test added. (Replied in-thread.)
- **`compile_call.go:277` (no `IsStable()` gate) — pushed back with evidence.** Verified the staleness is unreachable: a `define`-redefine resolves to an unstamped runtime binding (handled by the `cbIdx<0` path); `set!` on an import is rejected; both stamp seams only stamp non-rebindable bindings (`Imported ⇒ IsStable`). Documented the invariant instead of adding a dead branch. (Replied in-thread.)

### Crosscheck (code / errors / types / tests / consistency)

**Verified sound (no change needed):** templates are faithful transcriptions of the bootstrap clauses (code lens, verbatim-diffed — incl. `fold`'s `pair?` guard, evaluation order, edge cases); reclaim thresholds genuinely separate working/broken (errors+tests, *measured*: 0.0 vs 2.0, 4× margin); the string differential measurement is necessary and correct; the `any`-typed template-store boundary is the right layering and the map-of-strings is the right encoding (don't introduce an AST type); the parallel bench is structurally sound.

**Fixed:**
| Finding | Lens | Resolution |
|---|---|---|
| `*CapturingCallbackCorrect` tests vacuous (passive call/cc passed even under a forced false-positive) | tests (crit 9) | Replaced with **re-entrant** continuation tests for vector-map/string-map/fold — *verified they fail under the injected false-positive* |
| Two parallel maps must share keys, "single source of truth" comment false | types | Consolidated into one `inlineHOFSpecs{callbackParam, importGated, template}` |
| No build-time check that the stamped callback index is a valid template param | types | `validateInlineHOFCallbackIndex` (fail-closed at engine init, unit-tested) |
| Multi-arg arity fall-through unpinned | tests (7) | Equivalence cases added (`(vector-map + #(1 2 3) #(10 20 30))` → `#(11 22 33)`, etc.) |
| Empty/single-element inputs unpinned | tests (6) | Added to each correctness table |
| Missing hygiene tests for the new shapes | consistency / errors | `TestInlineHOF{VectorMap,StringMap,Fold}Hygiene` (shadow vector-ref/string-ref/car) |
| `binding.go` comment listed 5 HOFs, omitted `fold`; "equivalence" wording overstated | code / types | Corrected |

Not merging — awaiting your go-ahead.
